### PR TITLE
Do not use "ClusterIP:None" for Services (and add some tests)

### PR DESCRIPTION
**Pull request by chrisfarms**
## What

* `ClusterIP: None` is incompatible with the way Istio registers it's services, so we have altered the generated `Service` to get a propper `ClusterIP`
* As part of this work we also discovered that `Metadata` records could not have been updated (without destroying them first) and addressed this.
* The controller was poorly tested, so we did some refactoring to allow us to mock out the cloudhsm stuff and test the controller code without the need for an hsm wired up
* We spotted some issues where the Reconcile function would be called an unnecessarily large number of times, and have addressed this

### CODE REVIEW

File: pkg/controller/metadata/metadata_controller.go [hunk 1]
```diff
@@ -279,8 +281,18 @@ func (r *ReconcileMetadata) Reconcile(request reconcile.Request) (reconcile.Resu
...
+		updatedData, err := r.generateMetadataSecretData(instance, hsmCreds, hsmCreds)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		foundSecret.ObjectMeta.Annotations[versionAnnotation] = currentVersion
+		foundSecret.Data = updatedData
+		err = r.Update(context.TODO(), foundSecret)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to update Secret %s: %s", foundSecret.ObjectMeta.Name, err)
+		}
```
Comment (blairboy362) [inline pkg/controller/metadata/metadata_controller.go]: Could an `info` log message be added to indicate successful completion?

File: pkg/controller/metadata/metadata_controller.go [hunk 2]
```diff
@@ -348,14 +363,21 @@ func (r *ReconcileMetadata) Reconcile(request reconcile.Request) (reconcile.Resu
 		}
 	} else if err != nil {
 		return reconcile.Result{}, err
-	} else {
-		// TODO: Update deployment if changed
+	} else if foundDeployment.ObjectMeta.Annotations[versionAnnotation] != currentVersion {
+		log.Info("Updating Deployment", "namespace", metadataDeployment.Namespace, "name", metadataDeployment.Name)
+		err = r.Update(context.TODO(), foundDeployment)
+		if err != nil {
+			return reconcile.Result{}, err
```
Comment (blairboy362) [inline pkg/controller/metadata/metadata_controller.go]: Should this be wrapped with a "Failed to update deployment" as per Secret above?

File: pkg/controller/metadata/metadata_controller.go [hunk 3]
```diff
@@ -348,14 +363,21 @@ func (r *ReconcileMetadata) Reconcile(request reconcile.Request) (reconcile.Resu
...
+		}
```
Comment (blairboy362) [inline pkg/controller/metadata/metadata_controller.go]: Could an `info` log message be added to indicate successful completion?

File: pkg/controller/metadata/metadata_controller.go [hunk 4]
```diff
@@ -385,8 +406,24 @@ func (r *ReconcileMetadata) Reconcile(request reconcile.Request) (reconcile.Resu
...
+				Protocol:   "TCP",
+				Port:       80,
+				Name:       "http",
+				TargetPort: intstr.FromInt(80),
+			},
+		}
+
+		err = r.Update(context.TODO(), foundService)
+		if err != nil {
+			return reconcile.Result{}, err
```
Comment (blairboy362) [inline pkg/controller/metadata/metadata_controller.go]: Should this be wrapped in a "Failed to update Service" as per Secret above?

File: pkg/controller/metadata/metadata_controller.go [hunk 5]
```diff
@@ -385,8 +406,24 @@ func (r *ReconcileMetadata) Reconcile(request reconcile.Request) (reconcile.Resu
...
+		}
+
```
Comment (blairboy362) [inline pkg/controller/metadata/metadata_controller.go]: Could an `info` log message be added to indicate successful completion?

File: pkg/hsm/awscloudhsm/cloudhsm.go [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+		fmt.Sprintf("HSM_IP=%s", hsmCreds.IP),
+	)
+	cert, err := cmd.Output()
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate rsa key for %s: %s", label, err)
+	}
+	if !strings.Contains(string(cert), "--BEGIN CERTIFICATE--") {
+		return nil, fmt.Errorf("generated %s certificate does not appear to be a valid PEM format: %s", label, cert)
+	}
+	return cert, nil
```
Comment (blairboy362) [inline pkg/hsm/awscloudhsm/cloudhsm.go]: Could an `info` log message be added to indicate successful completion?

File: pkg/controller/metadata/metadata_controller.go (hunk 1 above)
Comment (chrisfarms) [reply pkg/controller/metadata/metadata_controller.go]: Sure ... Added a commit to pair up `creating-x` with `created-x` and `updating-x` with `updated-x` messages

File: pkg/controller/metadata/metadata_controller.go (hunk 2 above)
Comment (chrisfarms) [reply pkg/controller/metadata/metadata_controller.go]: 👍

File: pkg/controller/metadata/metadata_controller.go (hunk 5 above)
Comment (chrisfarms) [reply pkg/controller/metadata/metadata_controller.go]: yep

File: pkg/hsm/awscloudhsm/cloudhsm.go (hunk 6 above)
Comment (chrisfarms) [reply pkg/hsm/awscloudhsm/cloudhsm.go]: yep

File: pkg/controller/metadata/metadata_controller.go (hunk 4 above)
Comment (chrisfarms) [reply pkg/controller/metadata/metadata_controller.go]: done

**Pull request merged**